packages/msgs-derive/src/lib.rs

```diff
@@ -357,3 +457,274 @@ fn build_full_check_function(
+///     is sent by a proxy contract, the original sender is the address that initiated the transaction
+///     on the proxy.
+///
+/// This macro takes arguments of the form:
+///
+/// #\[external_execute(proxy(contract = find_contract_address), direct(sender = find_sender_address))\]
+///
+/// 'proxy' handles case 1, and 'direct' handles
+/// case 2. In both scenarios, the left hand side of the assignment is the identifier for the
+/// contract and original sender, respectively. The right hand side is a function with the signature:
+///
+/// FnOnce(&dyn cosmwasm_std::Storage) -> error_stack::Result<cosmwasm_std::Addr, impl error_stack::Context>
+///
+/// for contracts, and
+///
+/// FnOnce(&dyn cosmwasm_std::Storage, &ExecuteMsg) -> error_stack::Result<cosmwasm_std::Addr, impl error_stack::Context>
+///
+/// for addresses. The right hand side can be an expression that returns a function with that signature.
+#[proc_macro_attribute]
+pub fn external_execute(attr: TokenStream, item: TokenStream) -> TokenStream {
+    let mut execute_fn = syn::parse_macro_input!(item as ItemFn);
+    if execute_fn.sig.ident != format_ident!("execute") {
+        panic!("external_execute macro can only be used with execute endpoint")
+    }
+
+    let all_permissions = syn::parse_macro_input!(attr as AllPermissions);
+    let (contract_names, contract_permissions): (Vec<_>, Vec<_>) =
+        all_permissions.relay_permissions.into_iter().unzip();
+    let (_, specific_permissions): (Vec<_>, Vec<_>) = all_permissions
+        .specific_permissions
+        .into_iter()
+        .sorted_by(|a, b| sort_permissions(&a.0, &b.0))
+        .unzip();
+
+    let contract_names_literals: Vec<_> = contract_names
+        .iter()
+        .map(|cn| Literal::string(cn.to_string().as_str()))
+        .collect();
+
+    // Replace ExecuteMsg with ExecuteMsgFromProxy
+    let original_msg = execute_fn
+        .sig
+        .inputs
+        .pop()
+        .expect("error parsing execute endpoint's last argument")
+        .into_value();
+    let original_msg_ident = match original_msg {
+        syn::FnArg::Typed(typ) => match *typ.ty {
+            syn::Type::Path(p) => p
+                .path
+                .get_ident()
+                .expect("error parsing execute message type")
+                .clone(),
+            _ => panic!("problem parsing final argument of 'execute'"),
+        },
+        _ => panic!("last argument of 'execute' must be a typed execute message"),
+    };
+    let new_msg_ident = external_execute_msg_ident(original_msg_ident.clone());
+
+    execute_fn
+        .sig
+        .inputs
+        .push(parse_quote! {msg: #new_msg_ident});
+
+    let validate_fn = validate_external_contract_function(contract_names.clone());
+    let validate_fn = syn::parse_macro_input!(validate_fn as ItemFn);
+
+    let statements = execute_fn.block.stmts;
+    execute_fn.block = parse_quote!(
+        {
+            let (msg, info) = match msg {
```

Comment:
```suggestion
            let (msg, info) = match #new_msg_ident::from(msg) {
```
this should solve the `into` ambiguity

Reply:
I read the type annotation error incorrectly, this might not be needed (`msg.into()` should be enough)